commission/wizards/commission_make_settle.py

```diff
@@ -104,19 +104,20 @@ def action_settle(self):
         settlement_obj = self.env["commission.settlement"]
         settlement_line_obj = self.env["commission.settlement.line"]
         settlement_ids = []
-        if self.agent_ids:
-            agents = self.agent_ids
-        else:
-            agents = self.env["res.partner"].search([("agent", "=", True)])
+        agents = self.agent_ids or self.env["res.partner"].search(
+            [("agent", "=", True)]
+        )
         date_to = self.date_to
+        settlement_line_vals = []
         for agent in agents:
             date_to_agent = self._get_period_start(agent, date_to)
             # Get non settled elements
             agent_lines = self._get_agent_lines(agent, date_to_agent)
-            for company in agent_lines.mapped("company_id"):
-                agent_lines_company = agent_lines.filtered(
-                    lambda r: r.object_id.company_id == company
-                )
+            agent_lines_by_company = {}
+            for line in agent_lines:
+                agent_lines_by_company.setdefault(line.company_id, line)
+                agent_lines_by_company[line.company_id] |= line
```

Comment:
This set operation is slower than the filtered, so please keep the other.

Reply:
Do you know why this is slower? Because it's true that I didn't actually test the time differences until now and I can't really make the new solution better even in bigger `agent_lines` lengths (~1000) like I expected, with prefetching doing the bulk of the work already.

I'm open to change it back but I will explain that the idea behind this is that iterating once to group lines by company and once to iterate on the grouped lines being O(2n) sounds much better to me than the `filtered` that iterates through every line inside the `for`, making it O(n^n).
I saw a very similar case recently relating to a website method that drastically reduced the execution time from 60 to 20 seconds when having ~800 lines with a change nearly identical to this, which is why it seemed sensible to do

Reply:
The cache makes successive `filtered` operations to run smooth.

The problem with set operations is that the decision behind recordset operations is to copy the recordset structure, and in the copy, append the new record. This operation can be memory and time consuming over the other one. You will notice it with a bigger set.

Reply:
I see, I reverted it back to the original